pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// SPDX-License-Identifier: Apache-2.0
+
+package cas
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/good-night-oppie/helios-engine/pkg/helios/types"
+	"lukechampine.com/blake3"
+)
+
+// writeOp represents a background write operation
+type writeOp struct {
+	filePath string
+	content  []byte
+}
+
+// ContentAddressableStore defines the interface for content-addressable storage
+// implementing the core CAS operations with performance guarantees from research
+type ContentAddressableStore interface {
+	// Store saves content and returns its content-addressable hash
+	// Performance target: <1ms for typical code file sizes
+	Store(content []byte) (types.Hash, error)
+
+	// Load retrieves content by its hash
+	// Performance target: <5ms for typical code file sizes  
+	Load(hash types.Hash) ([]byte, error)
+
+	// Exists checks if content with given hash exists
+	// Performance target: <50μs for existence checks (relaxed during implementation)
+	Exists(hash types.Hash) bool
+
+	// Close releases resources
+	Close() error
+}
+
+// BLAKE3Store implements ContentAddressableStore using BLAKE3 hashing
+// Based on research findings showing BLAKE3's superior performance characteristics:
+// - ~15 GB/s throughput with AVX2
+// - <100ns latency for small inputs  
+// - SIMD acceleration support
+type BLAKE3Store struct {
+	storePath   string
+	cache       map[string][]byte // L1 cache for hot content
+	mutex       sync.RWMutex      // Protects cache for concurrent access
+	
+	// Performance optimizations from research phase
+	hasherPool  sync.Pool         // Pool of reusable BLAKE3 hashers
+	keyCache    map[string]string // Pre-computed hex keys for hot paths
+	keyMutex    sync.RWMutex      // Protects key cache
+	
+	// Ultra-performance optimizations for <70μs VST targets
+	memoryMode  bool              // Skip disk I/O for maximum performance
+	writeQueue  chan writeOp      // Async write queue for background persistence
+	errorQueue  chan error        // Channel for background write errors
+	wg          sync.WaitGroup    // Wait group for background writes
+	closed      bool              // Track if store is closed
+	closeMutex  sync.Mutex        // Protects close operations
+}
+
+// NewBLAKE3Store creates a new BLAKE3-based content-addressable store
+// storePath: directory for persistent storage
+func NewBLAKE3Store(storePath string) (*BLAKE3Store, error) {
+	// Ensure storage directory exists
+	if err := os.MkdirAll(storePath, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create storage directory: %w", err)
+	}
+
+	store := &BLAKE3Store{
+		storePath:  storePath,
+		cache:      make(map[string][]byte),
+		keyCache:   make(map[string]string),
+		memoryMode: false, // Default to persistent mode
+		writeQueue: make(chan writeOp, 1000), // Buffered channel for async writes
```

Comment:
Config: make writeQueue/errorQueue buffer sizes configurable (e.g., via options). Workloads vary; 1000/100 may be too small/large depending on environment.